plutus/lib/mysql.py

```diff
@@ -0,0 +1,171 @@
+        log.error("Unexpected error. Plutus should only manage budgets for single projects.")
+        metrics.incr("error_count", tags=["type:project_budget_filter", f"project_id:{project_id}",
+                                          f"config_type:{config_type}"])
+        sys.exit(1)
+
+    if 'services' in budget_dict['budget_filter']:
+        products = ','.join(budget_dict['budget_filter']['services'])
+    else:
+        products = "ALL"
+
+    owner_emails = get_owner_emails_for_project(project_id, alert_emails)
+
+    if budget_dict['amount'].get('specified_amount') is None:
+        budget_type = "LASTMONTH"
+        budget_amount = "-1"
+    else:
+        budget_type = "AMT"
+        budget_amount = budget_dict['amount']['specified_amount']['units']
+
+    if budget_dict['budget_filter']['credit_types_treatment'] == 'INCLUDE_ALL_CREDITS':
+        include_credits = "TRUE"
+    elif budget_dict['budget_filter']['credit_types_treatment'] == 'EXCLUDE_ALL_CREDITS':
+        include_credits = "FALSE"
+    else:
+        log.error("Unexpected error. Credits_type_treatment should be either of \
+                  INCLUDE_ALL_CREDITS or EXCLUDE_ALL_CREDITS.")
+        log.error("Google may have changed their budgets API.")
+        include_credits = "FALSE"
+        metrics.incr("error_count", tags=["type:credits_type_treatment",
+                                          f"project_id:{project_id}", f"config_type:{config_type}"])
+
+    if budget_dict['all_updates_rule'].get('pubsub_topic') is None:
+        pubsub = "FALSE"
+        pubsub_topic = "NA"
+    else:
+        pubsub = "TRUE"
+        pubsub_topic = budget_dict['all_updates_rule']['pubsub_topic']
+
+    now = datetime.datetime.utcnow()
+    curr_time = now.strftime('%Y-%m-%d %H:%M:%S')
+
+    last_modified = curr_time
+    created_date = curr_time
+
+    sql = f""" INSERT INTO budgets (
+budget_id,
+display_name,
+project_id,
+project_number,
+products,
+budget_type,
+budget_amount,
+include_credits,
+pubsub,
+pubsub_topic,
+owner_emails,
+created_date,
+last_modified,
+config_type
+)
+VALUES
+(
+'{budget_id}',
+'{display_name}',
+'{project_id}',
+'{project_number}',
+'{products}',
+'{budget_type}',
+{budget_amount},
+{include_credits},
+{pubsub},
+'{pubsub_topic}',
+'{owner_emails}',
+'{created_date}',
+'{last_modified}',
+'{config_type}'
+)
+ON DUPLICATE KEY UPDATE
+display_name = '{display_name}',
+project_id = '{project_id}',
+project_number = '{project_number}',
+products = '{products}',
+budget_type = '{budget_type}',
+budget_amount = {budget_amount},
+include_credits = {include_credits},
+pubsub = {pubsub},
+pubsub_topic = '{pubsub_topic}',
+owner_emails = '{owner_emails}',
+last_modified = '{last_modified}',
+config_type = '{config_type}'
+"""
+
+    try:
+        mysql_cursor.execute(sql)
+        metrics.incr("upsert_count", tags=[f"project_id:{project_id}",
+                                           f"config_type:{config_type}"])
+    except Exception as err:
```

Comment:
If this execution fails it should be fatal. Maybe `raise` after you log the error? Ideally you wouldn't have this catch all exception block.

Reply:
Changed to use specific exception types. As a design decision to always upsert, I don't want to stop the application if I encounter a db error (whether transient or not), since the error can be fixed later and state should normalize.